scrapy/pipelines/images.py

```diff
@@ -177,7 +185,16 @@ def _warn():
         ## end of deprecation warning block
 
         image_guid = hashlib.sha1(to_bytes(url)).hexdigest()  # change to request.url after deprecation
-        return 'full/%s.jpg' % (image_guid)
+        ext = "jpeg"
+        if self.convert_type == 'PRESERVE':
+            if response:
+                ext = Image.open(BytesIO(response.body)).format.lower()
+            else:
+                warnings.warn('You are preserving image type and no response body found.'
```

Comment:
Yes, the extension can be detected from the image type, I am using the format attribute and the lower casing it to get the extension. AFAICT from the PIL documentation this should work for all formats.